docs/source/basics/custom_components.rst

```diff
@@ -0,0 +1,171 @@
+    # Show all the default recipes
+    tune ls
+    # This makes a copy of the full finetune single device recipe locally
+    tune cp full_finetune_single_device ~/my_project/recipes/single_device.py
+    # Launch custom recipe with custom config from project directory
+    tune run recipes/single_device.py --config config/qwen_config.yaml
+
+If you are writing a new recipe from scratch, we recommend following the Python convention of defining a ``main()`` function
+in your script and decorating it with the :func:`~torchtune.config.parse` decorator. This will enable you to launch the recipe
+with ``tune run`` and pass in a yaml file for the ``--config`` argument.
+
+.. code-block:: python
+
+    from torchtune import config
+    from omegaconf import DictConfig
+
+    @config.parse
+    def main(cfg: DictConfig):
+        # Add all your recipe logic here, access config fields as attributes
+
+    if __name__ == "__main__":
+        # Config will be parsed from CLI, don't need to pass in here
+        main()
+
+Launching with custom components
+--------------------------------
+torchtune supports full experimentation with custom models, datasets, optimizers, or any fine-tuning component. You can define
+these locally in your repo and use them in your recipes and configs that you can launch with ``tune run``.
+
+We recommend following the "builder" pattern when making your components. This means creating "builder" functions that set up
+the classes you need with a few high level parameters that can be modified easily from the config. For example, we can define custom
+model and dataset builders in our project directory:
+
+.. code-block:: python
+
+    #
+    # In models/custom_decoder.py
+    #
+    class CustomTransformerDecoder(nn.Module):
+        # A custom architecture not present in torchtune
+
+    # Builder function for the custom model
+    def custom_model(num_layers: int, classification_head: bool = False):
+        # Any setup for defining the class
+        ...
+        # Return the module you want to train
+        return CustomTransformerDecoder(...)
+
+.. code-block:: python
+
+    #
+    # In datasets/custom_dataset.py
+    #
+    from torchtune.datasets import SFTDataset, PackedDataset
+    from torchtune.data import InputOutputToMessages
+    from torchtune.modules.tokenizers import ModelTokenizer
+
+    # Example builder function for a custom code instruct dataset not in torchtune, but using
+    # different dataset building blocks from torchtune
+    def tiny_codes(tokenizer: ModelTokenizer, packed: bool = True):
+        """
+        Python subset of nampdn-ai/tiny-codes. Instruct and code response pairs.
+        """
+        ds = SFTDataset(
+            model_transform=tokenizer,
+            source="nampdn-ai/tiny-codes",
+            message_transform=InputOutputToMessages(
+                column_map={"input": "prompt", "output": "response"},
+            ),
+            filter_fn=lambda x: x["language"] == "python",
```

Comment:
this is a really slick example

Reply:
I actually used this too so it's ~~authentic~~